src/spec.c

```diff
@@ -235,47 +235,124 @@ RSValueType fieldTypeToValueType(FieldType ft) {
       return RSValue_Null;
   }
 }
-void _spec_buildSortingTable(IndexSpec *spec, int len) {
-  spec->sortables = NewSortingTable(len);
-  for (int i = 0; i < spec->numFields; i++) {
-    if (FieldSpec_IsSortable(&spec->fields[i])) {
-      // printf("Adding sortable field %s id %d\n", spec->fields[i].name, spec->fields[i].sortIdx);
-      SortingTable_SetFieldName(spec->sortables, spec->fields[i].sortIdx, spec->fields[i].name,
-                                fieldTypeToValueType(spec->fields[i].type));
+
+static int IndexSpec_AddFieldsInternal(IndexSpec *sp, const char **argv, int argc, char **err,
+                                       int isNew) {
+
+  int textId = -1;
+  for (size_t ii = 0; ii < sp->numFields; ++ii) {
+    const FieldSpec *fs = sp->fields + ii;
+    if (fs->type == FIELD_FULLTEXT) {
+      textId = MAX(textId, fs->textOpts.id);
     }
   }
+
+  for (int offset = 0; offset < argc && sp->numFields < SPEC_MAX_FIELDS;) {
+    sp->fields = rm_realloc(sp->fields, sizeof(*sp->fields) * (sp->numFields + 1));
+    FieldSpec *fs = sp->fields + sp->numFields;
+    memset(fs, 0, sizeof(*fs));
+
+    fs->index = sp->numFields;
+
+    if (!parseFieldSpec(argv, &offset, argc, fs, err)) {
+      if (!*err) {
+        SET_ERR(err, "Could not parse field spec");
+      }
+      return 0;
+    }
+
+    if (fs->type == FIELD_FULLTEXT && FieldSpec_IsIndexable(fs)) {
+      // make sure we don't have too many indexable fields
+      textId++;  // Explicit
+      if (textId == SPEC_MAX_FIELD_ID) {
+        SET_ERR(err, "Too many TEXT fields in schema");
+        return 0;
+      }
+
+      // If we need to store field flags and we have over 32 fields, we need to switch to wide
+      // schema encoding
+      if (textId >= SPEC_WIDEFIELD_THRESHOLD && (sp->flags & Index_StoreFieldFlags)) {
+        if (isNew) {
+          sp->flags |= Index_WideSchema;
+        } else if ((sp->flags & Index_WideSchema) == 0) {
+          SET_ERR(err,
+                  "Cannot add more fields. Declare index with wide fields to allow adding "
+                  "unlimited fields");
+          return 0;
+        }
+      }
+      fs->textOpts.id = textId;
+    }
+
+    if (IndexSpec_GetField(sp, fs->name, strlen(fs->name))) {
+      SET_ERR(err, "Duplicate field in schema");
+      return 0;
+    }
+
+    if (FieldSpec_IsSortable(fs)) {
+      fs->sortIdx = RSSortingTable_Add(sp->sortables, fs->name, fieldTypeToValueType(fs->type));
+    }
+    sp->numFields++;
+  }
+  return 1;
+}
+
+int IndexSpec_AddFields(IndexSpec *sp, const char **argv, int argc, char **err) {
+  // Store the old field count and sortable size so that if validation fails, the
+  // index is still usable.
+  size_t oldFieldCount = sp->numFields;
+  size_t oldSortLen = sp->sortables->len;
+  if (!IndexSpec_AddFieldsInternal(sp, argv, argc, err, 0)) {
+    sp->numFields = oldFieldCount;
+    sp->sortables->len = oldSortLen;
```

Comment:
So its possible that the "IndexSpec_AddFieldsInternal" will fail but will increase the "numFields" field and the sortables->len? if so I think "IndexSpec_AddFieldsInternal" should handle the correctness of those values.

What do you think?

Reply:
I see what you mean, but with cases like these, it's usually easier to have a wrapper function (like I have here) rather than encase the function body in a giant if loop. Let me comment in the function though, that the index size etc. may become invalid.

Reply:
Actually, I'll change it. You are right that we might make some more changes, and it's easier to keep track of this in one function than it is in two.